pkg/cuectx/load.go

```diff
@@ -0,0 +1,226 @@
+//
+// TODO this approach is complicated and confusing, refactor to something understandable
+func LoadGrafanaInstancesWithThema(path string, cueFS fs.FS, rt *thema.Runtime, opts ...thema.BindOption) (thema.Lineage, error) {
+	prefix := filepath.FromSlash(path)
+	fs, err := prefixWithGrafanaCUE(prefix, cueFS)
+	if err != nil {
+		return nil, err
+	}
+	inst, err := load.InstanceWithThema(fs, prefix)
+
+	// Need to trick loading by creating the embedded file and
+	// making it look like a module in the root dir.
+	if err != nil {
+		return nil, err
+	}
+
+	val := rt.Context().BuildInstance(inst)
+
+	lin, err := thema.BindLineage(val, rt, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return lin, nil
+}
+
+// prefixWithGrafanaCUE constructs an fs.FS that merges the provided fs.FS with
+// the embedded FS containing Grafana's core CUE files, [grafana.CueSchemaFS].
+// The provided prefix should be the relative path from the grafana repository
+// root to the directory root of the provided inputfs.
+//
+// The returned fs.FS is suitable for passing to a CUE loader, such as [load.InstanceWithThema].
+func prefixWithGrafanaCUE(prefix string, inputfs fs.FS) (fs.FS, error) {
+	m, err := prefixFS(prefix, inputfs)
+	if err != nil {
+		return nil, err
+	}
+	return merged_fs.NewMergedFS(m, grafana.CueSchemaFS), nil
+}
+
+// TODO such a waste, replace with stateless impl that just transforms paths on the fly
+func prefixFS(prefix string, fsys fs.FS) (fs.FS, error) {
+	m := make(fstest.MapFS)
+
+	prefix = filepath.FromSlash(prefix)
+	err := fs.WalkDir(fsys, ".", func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if d.IsDir() {
+			return nil
+		}
+
+		b, err := fs.ReadFile(fsys, filepath.ToSlash(path))
+		if err != nil {
+			return err
+		}
+		// fstest can recognize only forward slashes.
+		m[filepath.ToSlash(filepath.Join(prefix, path))] = &fstest.MapFile{Data: b}
+		return nil
+	})
+	return m, err
+}
+
+// LoadGrafanaInstance wraps [load.InstanceWithThema] to load a
+// [*build.Instance] corresponding to a particular path within the
+// github.com/grafana/grafana CUE module.
+//
+// This allows resolution of imports within the grafana or thema CUE modules to
+// work correctly and consistently by relying on the embedded FS at
+// [grafana.CueSchemaFS] and [thema.CueFS].
+//
+// relpath should be a relative path path within [grafana.CueSchemaFS] to be
+// loaded. Optionally, the caller may provide an additional fs.FS via the
+// overlay parameter, which will be merged with [grafana.CueSchemaFS] at
+// relpath, and loaded.
+//
+// pkg, if non-empty, is set as the value of
+// ["cuelang.org/go/cue/load".Config.Package]. If the CUE package to be loaded
+// is the same as the parent directory name, it should be omitted.
+//
+// NOTE this function will be deprecated in favor of a more generic loader
```

Comment:
Will the more generic loader remain in pkg/cuectx? I am more generally wondering about how to decide what should go to `kindsys` vs `pkg/cuectx`. I'm asking out of curiosity because I'm trying to learn more about this topic.

Reply:
Good question - I plan on adding a generic loader function in kindsys and use that to replace the grafana-specific loader (the generic loader will be also be used in the SDK).